File: examples/go/smtp_client.go
// Package mail provides a minimal SMTP client for sending mail through
// the uye-mail server from Go microservices.
//
// In dev:  connect to mailserver:587 (or mailpit:1025 directly)
// In prod: connect to mailserver:587 with STARTTLS
package mail

import (
	"crypto/tls"
	"fmt"
	"net/smtp"
	"strings"
)

// Config holds SMTP connection parameters.
// Populate from environment variables in your service (e.g. via os.Getenv).
type Config struct {
	// Host is the SMTP server hostname — use the Docker service name
	// ("mailserver") when running in the same Docker network.
	Host string

	// Port is typically 587 (STARTTLS submission).
	Port int

	// Username and Password are the credentials of the sending mailbox.
	Username string
	Password string

	// From is the envelope sender address.
	From string

	// InsecureSkipVerify disables TLS certificate verification.
	// Set to true in dev when using self-signed certificates.
	InsecureSkipVerify bool
}

// DefaultDevConfig returns a Config pointing at the mailserver Docker service
// with settings suitable for local development.
func DefaultDevConfig() Config {
	return Config{
		Host:               "mailserver",
		Port:               587,
		Username:           "noreply@example.com",
		Password:           "changeme",
		From:               "noreply@example.com",
		InsecureSkipVerify: true, // self-signed cert in dev
	}
}

// Message represents an outbound email.
type Message struct {
	To      []string
	Subject string
	Body    string // plain text
}

// Send delivers msg via the configured SMTP server using STARTTLS.
func Send(cfg Config, msg Message) error {
	if len(msg.To) == 0 {
		return fmt.Errorf("mail: no recipients")
	}

	auth := smtp.PlainAuth("", cfg.Username, cfg.Password, cfg.Host)
	addr := fmt.Sprintf("%s:%d", cfg.Host, cfg.Port)

	tlsCfg := &tls.Config{
		ServerName:         cfg.Host,
		InsecureSkipVerify: cfg.InsecureSkipVerify, //nolint:gosec
	}

	// Dial a plain TCP connection first, then upgrade to TLS via STARTTLS.
	conn, err := tls.Dial("tcp", addr, tlsCfg)
	if err != nil {
		// Fall back to plain + STARTTLS if direct TLS dial fails
		// (port 587 uses STARTTLS, not implicit TLS)
		return sendSTARTTLS(addr, auth, cfg, msg, tlsCfg)
	}
	defer conn.Close()

	client, err := smtp.NewClient(conn, cfg.Host)
	if err != nil {
		return fmt.Errorf("mail: smtp client: %w", err)
	}
	defer client.Close()

	return deliver(client, auth, cfg.From, msg)
}

func sendSTARTTLS(addr string, auth smtp.Auth, cfg Config, msg Message, tlsCfg *tls.Config) error {
	client, err := smtp.Dial(addr)
	if err != nil {
		return fmt.Errorf("mail: dial %s: %w", addr, err)
	}
	defer client.Close()

	if err := client.StartTLS(tlsCfg); err != nil {
		return fmt.Errorf("mail: starttls: %w", err)
	}

	return deliver(client, auth, cfg.From, msg)
}

func deliver(client *smtp.Client, auth smtp.Auth, from string, msg Message) error {
	if err := client.Auth(auth); err != nil {
		return fmt.Errorf("mail: auth: %w", err)
	}
	if err := client.Mail(from); err != nil {
		return fmt.Errorf("mail: MAIL FROM: %w", err)
	}
	for _, to := range msg.To {
		if err := client.Rcpt(to); err != nil {
			return fmt.Errorf("mail: RCPT TO %s: %w", to, err)
		}
	}

	w, err := client.Data()
	if err != nil {
		return fmt.Errorf("mail: DATA: %w", err)
	}
	defer w.Close()

	raw := buildRaw(from, msg)
	if _, err := fmt.Fprint(w, raw); err != nil {
		return fmt.Errorf("mail: write body: %w", err)
	}
	return nil
}

func buildRaw(from string, msg Message) string {
	return strings.Join([]string{
		"From: " + from,
		"To: " + strings.Join(msg.To, ", "),
		"Subject: " + msg.Subject,
		"MIME-Version: 1.0",
		"Content-Type: text/plain; charset=UTF-8",
		"",
		msg.Body,
	}, "\r\n")
}

File: examples/python/smtp_client.py
"""
smtp_client.py — Minimal SMTP client for sending mail from Python microservices
through the uye-mail server.

In dev:  connect to mailserver:587 (or mailpit:1025 directly)
In prod: connect to mailserver:587 with STARTTLS

Usage:
    from smtp_client import MailConfig, Message, send

    cfg = MailConfig.from_env()
    send(cfg, Message(
        to=["user@example.com"],
        subject="Hello",
        body="World",
    ))
"""

from __future__ import annotations

import os
import smtplib
import ssl
from dataclasses import dataclass
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from typing import List, Optional


@dataclass
class MailConfig:
    # Docker service name — "mailserver" when in the same Docker network.
    host: str = "mailserver"
    port: int = 587
    username: str = ""
    password: str = ""
    from_addr: str = ""
    # Set to True in dev to skip TLS cert verification (self-signed certs).
    insecure_skip_verify: bool = False

    @classmethod
    def from_env(cls) -> "MailConfig":
        """Load config from environment variables (12-factor style)."""
        return cls(
            host=os.environ.get("SMTP_HOST", "mailserver"),
            port=int(os.environ.get("SMTP_PORT", "587")),
            username=os.environ.get("SMTP_USERNAME", ""),
            password=os.environ.get("SMTP_PASSWORD", ""),
            from_addr=os.environ.get("SMTP_FROM", ""),
            insecure_skip_verify=os.environ.get("SMTP_INSECURE", "").lower() == "true",
        )

    @classmethod
    def dev(cls) -> "MailConfig":
        """Convenience preset for local development."""
        return cls(
            host="mailserver",
            port=587,
            username="noreply@example.com",
            password="changeme",
            from_addr="noreply@example.com",
            insecure_skip_verify=True,
        )


@dataclass
class Message:
    to: List[str]
    subject: str
    body: str                        # plain text body
    html_body: Optional[str] = None  # optional HTML alternative


def send(config: MailConfig, message: Message) -> None:
    """
    Send an email via STARTTLS on port 587.
    Raises smtplib.SMTPException on failure.
    """
    if not message.to:
        raise ValueError("No recipients specified")

    msg = _build_mime(config.from_addr, message)

    ctx = ssl.create_default_context()
    if config.insecure_skip_verify:
        ctx.check_hostname = False
        ctx.verify_mode = ssl.CERT_NONE

    with smtplib.SMTP(config.host, config.port, timeout=10) as server:
        server.ehlo()
        server.starttls(context=ctx)
        server.ehlo()
        if config.username:
            server.login(config.username, config.password)
        server.sendmail(config.from_addr, message.to, msg.as_string())


def _build_mime(from_addr: str, message: Message) -> MIMEMultipart | MIMEText:
    if message.html_body:
        msg: MIMEMultipart | MIMEText = MIMEMultipart("alternative")
        msg.attach(MIMEText(message.body, "plain", "utf-8"))
        msg.attach(MIMEText(message.html_body, "html", "utf-8"))
    else:
        msg = MIMEText(message.body, "plain", "utf-8")

    msg["From"] = from_addr
    msg["To"] = ", ".join(message.to)
    msg["Subject"] = message.subject
    return msg
